cpp/src/arrow/engine/expression.h

```diff
@@ -0,0 +1,490 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "arrow/engine/catalog.h"
+#include "arrow/engine/type_fwd.h"
+#include "arrow/result.h"
+#include "arrow/type.h"
+#include "arrow/type_fwd.h"
+#include "arrow/util/compare.h"
+#include "arrow/util/variant.h"
+
+namespace arrow {
+namespace engine {
+
+/// ExprType is a class representing the type of an Expression. The type is
+/// composed of a shape and a DataType or a Schema depending on the shape.
+///
+/// ExprType is mainly used to validate arguments for operator expressions, e.g.
+/// relational operator expressions expect inputs of Table shape.
+///
+/// The sum-type representation would be:
+///
+/// enum ExprType {
+///   ScalarType(DataType),
+///   ArrayType(DataType),
+///   TableType(Schema),
+/// }
+class ARROW_EXPORT ExprType : public util::EqualityComparable<ExprType> {
+ public:
+  enum Shape : uint8_t {
+    // The expression yields a Scalar, e.g. "1".
+    SCALAR,
+    // The expression yields an Array, e.g. "[1, 2, 3]".
+    ARRAY,
```

Comment:
One unknown here is I don't know if we need to differentiate from Array and Column. For example, a user can still pass an "inline" array. Some operators (e.g. FilterRelExpr) requires that the predicate/mask has the same number of elements than the input table.

Reply:
Is there an operator which doesn't require this? For example, I think `In`'s argument will be `ExprType::Scalar|Array(list(some_type()))` rather than a top level array

Reply:
What the difference between Array and Column, the latter one is named, right?
I don't think we need to differentiate because their shape is identical.

Reply:
@kszucs no the main difference would be an implicit length constraint; an expression of Column shape would have equal length to all sibling expressions of Column shape.

@fsaintjacques what cases do you have in mind where an Array would have differing length from a Column?

Reply:
I think the main difference is knowledge of the length of the array. Column has an implicit that the length is the same as the referenced table. Where a table may or not have this.

An example, a user could have computed the mask out-of-band in panda because we're missing a functionality. In this case, from the perspective of the LogicalPlan, this is a value array that does not come from an expression of a column. It needs to validate the length. The same goes for any operators, e.g. in projection I could provide an array literal to concatenate to an existing table as a new column.

Reply:
The case you raise is interesting, but sounds more like something we would support through a custom `Expr` subclass than through a (guessing) first class `ArrayExpr`.

WRT the implicit length assumption: we'll still need to insert validation for this on translation to a physical plan, since a non-custom expression could still result in Columns of differing length:

```
# example 1:
# no custom, but one column is filtered while the other isn't
# so they probably won't have the same length
And(
  Filter(FieldRef("int32_column"), FieldRef("mask")),
  FieldRef("bool_column"))

# example 2:
# custom; no guarantee that my_pandas_mask has equal size to int32_column
# physical plan must include a length check
Filter(FieldRef("int32_column"), CustomArrayExpr(my_pandas_mask))
```

The physical plan builder should recognize when the column length assumption is not guaranteed by an expression tree and insert length validation nodes. (Example 1 is a toy; rather than inserting a length validation node we should probably raise an error on translation to a physical plan since for non trivial `mask` it *must* break the column length assumption.)